### ISSUE TITLE
encounter a problem while train with real hdr data

### ISSUE THREAD

**Issue by PK15946** [gabrieleilertsen/hdrcnn#14]
Hi again @gabrieleilertsen 
I am working on a pytorch implementation of your hdrcnn. I follow your instructions in the paper, which is load a pretrained vgg_places365 model and then train the network with places365 dataset. The result is fine. As shown below

Input
![input](https://user-images.githubusercontent.com/34185974/45020920-cc8ee300-b062-11e8-8536-833259d81985.png)

Output
![predict](https://user-images.githubusercontent.com/34185974/45020973-f0522900-b062-11e8-8950-b6e0af990c08.png)

However, When I load the pretrained model and train with real hdr data, unexpectedly, the result was degraded. Here is a 1-million-steps result.
![predict](https://user-images.githubusercontent.com/34185974/45021759-fa752700-b064-11e8-8879-f9a2309e7d9c.png)

Do you have such phenomenon when train with real hdr data?

**Comment by gabrieleilertsen** [comment]
Is the test image you show JPEG compressed? And do you include compression in the training data? The artifacts look similar to those encountered when running inference on a JPEG compressed image using a model that is trained without compressed images...

**Comment by PK15946** [comment]
The -jpeg_quality for generating train and pretrain data is set to 100, and the test image is in .tif format, could this be a problem? But I test the same image using your model with your ORIGINAL parameters. Here is the result. It looks much better.
![000007_out](https://user-images.githubusercontent.com/34185974/45146184-f37d1e80-b1f4-11e8-8d9d-c338b9fc0b03.png)

**Comment by gabrieleilertsen** [comment]
Ok, then it seems to be some other problem. Have you tried to first run the tensorflow training script to see that you get the expected result? If so, there should be some difference in your pytorch implementation that causes the problem.

If you get the same problem with the tensorflow training, there is possibly some problem with the HDR training data. How many HDR images are you using?

**Comment by PK15946** [comment]
The result above is trained with 688 hdr images( about 13000 after preprocessed). I suspect that it is too small, so I am collecting more data according to your list. But I can't find the HDR BOOK dataset, since it contains considerable amount images, could you please tell me where can I find those data?

**Comment by gabrieleilertsen** [comment]
Yes, sounds like the limited number of HDR images could be the problem.

The images from the 2005 HDR book are not available online. A great online source of HDR are the [HDR videos by Fröhlich et al.](https://hdr-2014.hdm-stuttgart.de/), which makes for many images if you choose e.g. every 10th frame or similar.

**Comment by PK15946** [comment]
Hello, @gabrieleilertsen 

I download the hdrcnn-master.zip and unzip it, then completely follow your scripts(except the compilation, which I use  "gcc virtualcamera.cpp -o virtualcamera -lm -lstdc++ -lopencv_core -lopencv_imgproc -lopencv_imgcodecs -std=c++11"), but it seems that the net was not trained correctly, Train loss and valid loss were both nan. 

So I check the training data, but it looks fine to me, as shown below.

![im_000003_000001](https://user-images.githubusercontent.com/34185974/45610611-22b24c00-ba8f-11e8-98a3-cbe23f57f8ef.jpg)
![im_000004_000001](https://user-images.githubusercontent.com/34185974/45610612-22b24c00-ba8f-11e8-8f9d-6ff874ad680b.jpg)
![im_000001_000001](https://user-images.githubusercontent.com/34185974/45610608-2219b580-ba8f-11e8-90cd-0e5b41a5e5a3.jpg)
![im_000002_000001](https://user-images.githubusercontent.com/34185974/45610610-22b24c00-ba8f-11e8-9469-5cdb1c10e4c2.jpg)

**Comment by gabrieleilertsen** [comment]
Very difficult to say what the problem can be without more information. My initial thought would be that there are pixel values <= 0 in the training data, or pixels that are NaN. The data preparation should take care of negative values (see rows [304-305](https://github.com/gabrieleilertsen/hdrcnn/blob/ae7d22146299ea3776aa6ad1edca0ad80743f5b5/training_code/virtualcamera/virtualcamera.cpp#L304) in the code). However, maybe something happened after that point, so to be sure you could check for this in the training images.

**Comment by PK15946** [comment]
Thanks for kindly helping me in this matter, @gabrieleilertsen 
I check my dataset, there is no pixel value <0 or = nan in those .bin files and .jpg files. But still, the train loss is nan. 
I mean no offense, could you please have a try with files downloaded from github?

**Comment by PK15946** [comment]
Thanks for kindly helping me in this matter, @gabrieleilertsen
I check my dataset, there is no pixel value <0 or = nan in those .bin files and .jpg files. But still, the train loss is nan.
I mean no offense, can you try the code downloaded from github?

**Comment by gabrieleilertsen** [comment]
The training script is tested before I uploaded it. There must be some problem either with the training data, or due to different installation environment. Maybe you could start by testing on a very small number of HDR images...

**Comment by PK15946** [comment]
I use 2 images, one for training, the other one for validating. But I still get nan loss. Can you upload 5 training pairs(.bin and .jpg) so I can determine the problem?

**Comment by gabrieleilertsen** [comment]
Sorry for the late reply. I hope you were able to find the problem.

I was going to look at this a few days ago, but I had updated my Cuda version so that the OpenCV Cuda support did not work. However, when recompiling OpenCV I run in to problems. Apparently, there is some problems in compiling OpenCV with Cuda 9.0 and I did not have time to struggle with this. Consequently, I cannot run the virtual camera application to generate training data at the moment. I can get back on this problem when I manage to find time, if there still is a problem for you to get the training running.

**Comment by PK15946** [comment]
Thanks in advance!
I am not familiar with tensorflow, but I learned to debug these days, the tfdbg told me the first place that inf value occur was in 'encoder/h5/conv_1/Conv2d:0, I am not sure whether it is the training data problem  or not, because if it is, I guess the inf/nan value should occur in the first layer.
I don't want this problem impede your current research, so I will keep trying, and your help is welcomed when you are free.
![selection_005](https://user-images.githubusercontent.com/34185974/46594591-b6af8a80-cb06-11e8-95e0-3f0c444894a3.png)
![selection_006](https://user-images.githubusercontent.com/34185974/46594592-b6af8a80-cb06-11e8-9679-23568bf4b93e.png)

**Comment by gabrieleilertsen** [comment]
Was able to test it now, and I cannot see any problems. I uploaded some images [here](http://hdrv.org/hdrcnn/material/test_data.zip), so you can test if you encounter the same problem with this data. There are 5 original exr images, which are processed to generate 120 binary and jpg training pairs.

You can try both using original images, and process them with the virtual camera application:
```
python3 hdrcnn_train.py --raw_dir=test_data/original --vgg_path=vgg16_places365_weights.npy --preprocess=1
```

Or you can run with the provided processed images:
```
python3 hdrcnn_train.py --data_dir=test_data/training_data --vgg_path=vgg16_places365_weights.npy --preprocess=0
```

You may have to reduce the batch size depending on available memory. The VGG initialization weights are available [here](http://hdrv.org/hdrcnn/material/vgg16_places365_weights.npy).

I hope this can help you find your problem. If you can run the above trainings without problems, there is something wrong in your training data. If you cannot, the problem is due do differences in our training environments, which would be good to find out what it is.

**Comment by PK15946** [comment]
Those data really help me a lot!
The Nan-problem was located in the  VGG-weight initialization file I have downloaded before. I never expect it was corrupted because when I load the that file, it looks so normal, without warning or error.

**Comment by ghost** [comment]
@PK15946 你好，可以把vgg16_places365_weights.npy文件发给我一份吗，作者给的链接下载不了。zhuixun10@foxmail.com，谢谢！

**Comment by wbhu** [comment]
> Those data really help me a lot!
> The Nan-problem was located in the VGG-weight initialization file I have downloaded before. I never expect it was corrupted because when I load the that file, it looks so normal, without warning or error.

Hi @PK15946,

I also want to implement this work by PyTorch. Could you please share your code. Thanks very much.

**Comment by weiwang90** [comment]
> Those data really help me a lot!
> The Nan-problem was located in the VGG-weight initialization file I have downloaded before. I never expect it was corrupted because when I load the that file, it looks so normal, without warning or error.

Hi，Can you share the vgg16_places365_weights.npy to me? The link website is down.
My Email is 1300027@wust.edu.cn
